helix-core/src/movement.rs

```diff
@@ -43,6 +43,47 @@ pub fn move_horizontally(
     range.put_cursor(slice, new_pos, behaviour == Movement::Extend)
 }
 
+pub fn move_horizontally_no_wrap(
+    slice: RopeSlice,
+    range: Range,
+    dir: Direction,
+    count: usize,
+    behaviour: Movement,
+) -> Range {
+    let pos = range.cursor(slice);
+
+    // Compute the new position.
+    let mut new_pos = match dir {
+        Direction::Forward => nth_next_grapheme_boundary(slice, pos, count),
+        Direction::Backward => nth_prev_grapheme_boundary(slice, pos, count),
+    };
+
+    new_pos = limit_to_line(slice, pos, new_pos, dir);
+
+    // Compute the final new range.
+    range.put_cursor(slice, new_pos, behaviour == Movement::Extend)
+}
+
+fn limit_to_line(slice: RopeSlice, old_pos: usize, new_pos: usize, dir: Direction) -> usize {
+    match dir {
+        Direction::Forward => {
+            if let Some(next_new_line) =
+                find_next_in_range(slice, Range::new(old_pos, new_pos), char_is_line_ending)
+            {
+                return new_pos.min(next_new_line);
+            }
+        }
+        Direction::Backward => {
+            if let Some(prev_new_line) =
+                find_prev_in_range(slice, Range::new(new_pos, old_pos), char_is_line_ending)
+            {
+                return new_pos.max(prev_new_line + 1);
+            }
+        }
+    }
+    new_pos
+}
```

Comment:
This is very expensive because it uses grapheme traversal. Let's instead use

```rust
let line = slice.char_to_line(original_pos);
let start = slice.line_to_char(line);
let end = slice.line_to_char(line + 1);

pos.clamp(start, end)
```